[abseil-cpp] Cherry-pick of https://github.com/abseil/abseil-cpp/pull/1801

We can drop this patch once we upgrade to a newer commit that contains it.

From: Jeremy Nimmer <jeremy.nimmer@tri.global>
Date: Mon, 30 Dec 2024 17:26:14 -0800
Subject: [PATCH] Remove unncessary <iostream> includes

Including iostream means introducing the static (global) constructors
and destructors for std::cin, std::cerr, and std::cout. That extra
init and fini code is undesirable when those streams are not actually
used.

--- absl/random/internal/randen_engine.h
+++ absl/random/internal/randen_engine.h
@@ -18,9 +18,10 @@
 #include <algorithm>
 #include <cinttypes>
 #include <cstdlib>
-#include <iostream>
+#include <istream>
 #include <iterator>
 #include <limits>
+#include <ostream>
 #include <type_traits>
 
 #include "absl/base/internal/endian.h"

--- absl/random/seed_gen_exception.cc
+++ absl/random/seed_gen_exception.cc
@@ -14,10 +14,12 @@
 
 #include "absl/random/seed_gen_exception.h"
 
-#include <iostream>
-
 #include "absl/base/config.h"
 
+#ifndef ABSL_HAVE_EXCEPTIONS
+#include <iostream>
+#endif
+
 namespace absl {
 ABSL_NAMESPACE_BEGIN
 

--- absl/strings/cord.cc
+++ absl/strings/cord.cc
@@ -23,7 +23,6 @@
 #include <cstring>
 #include <iomanip>
 #include <ios>
-#include <iostream>
 #include <limits>
 #include <memory>
 #include <ostream>

--- absl/strings/internal/charconv_bigint.h
+++ absl/strings/internal/charconv_bigint.h
@@ -17,7 +17,7 @@
 
 #include <algorithm>
 #include <cstdint>
-#include <iostream>
+#include <ostream>
 #include <string>
 
 #include "absl/base/config.h"

--- absl/strings/internal/cord_rep_btree.cc
+++ absl/strings/internal/cord_rep_btree.cc
@@ -17,10 +17,13 @@
 #include <atomic>
 #include <cassert>
 #include <cstdint>
-#include <iostream>
 #include <ostream>
 #include <string>
 
+#ifndef NDEBUG
+#include <iostream>
+#endif
+
 #include "absl/base/attributes.h"
 #include "absl/base/config.h"
 #include "absl/base/internal/raw_logging.h"

--- absl/strings/internal/cord_rep_btree_navigator.h
+++ absl/strings/internal/cord_rep_btree_navigator.h
@@ -16,7 +16,6 @@
 #define ABSL_STRINGS_INTERNAL_CORD_REP_BTREE_NAVIGATOR_H_
 
 #include <cassert>
-#include <iostream>
 
 #include "absl/strings/internal/cord_internal.h"
 #include "absl/strings/internal/cord_rep_btree.h"
